File: test/js/last-performance.test.js
// Unit tests for the pure "previous performance" lookup (coach/last-performance.js).
// Matching is by canonical_slug across ANY past workout, returning the most
// recent prior session that was actually logged with set data. An optional
// exposure narrows the chain to one recurring exposure of that movement, read
// from each date's historical plan.
import test from 'node:test';
import assert from 'node:assert/strict';
import {
    findLastPerformance,
    setHasData,
    formatShortDate,
} from '../../public/js/coach/last-performance.js';

// Helpers to build plan/log fixtures keyed by date. `exposure` is omitted from
// the exercise entirely when not given — the wire shape for the common case.
const planWith = (exId, slug, exposure) => ({
    blocks: [{ exercises: [{ id: exId, canonical_slug: slug, ...(exposure ? { exposure } : {}) }] }],
});
const logWith = (exId, sets) => ({ [exId]: { sets } });

test('setHasData: true only when a real metric is present', () => {
    assert.equal(setHasData(null), false);
    assert.equal(setHasData({}), false);
    assert.equal(setHasData({ set_num: 1 }), false);
    assert.equal(setHasData({ completed: true }), false);
    assert.equal(setHasData({ weight: 0 }), true);   // 0 is real data
    assert.equal(setHasData({ reps: 5 }), true);
    assert.equal(setHasData({ rpe: 8 }), true);
    assert.equal(setHasData({ duration_sec: 40 }), true);
});

test('no history -> null', () => {
    assert.equal(findLastPerformance('squat', '2026-06-07', {}, {}), null);
});

test('missing/empty slug -> null', () => {
    const plans = { '2026-06-01': planWith('ex_1', 'squat') };
    const logs = { '2026-06-01': logWith('ex_1', [{ set_num: 1, reps: 5 }]) };
    assert.equal(findLastPerformance('', '2026-06-07', plans, logs), null);
    assert.equal(findLastPerformance(null, '2026-06-07', plans, logs), null);
});

test('one prior in a different workout (different exercise_key) -> that session', () => {
    // The slug is "squat" but the per-session key differs across workouts.
    const plans = { '2026-06-01': planWith('ex_3', 'squat') };
    const logs = { '2026-06-01': logWith('ex_3', [{ set_num: 1, weight: 60, reps: 8, rpe: 8 }]) };
    const r = findLastPerformance('squat', '2026-06-07', plans, logs);
    assert.equal(r.date, '2026-06-01');
    assert.deepEqual(r.sets, [{ set_num: 1, weight: 60, reps: 8, rpe: 8 }]);
});

test('multiple priors -> most recent chosen', () => {
    const plans = {
        '2026-05-20': planWith('ex_1', 'squat'),
        '2026-06-03': planWith('ex_9', 'squat'),
    };
    const logs = {
        '2026-05-20': logWith('ex_1', [{ set_num: 1, weight: 50, reps: 10 }]),
        '2026-06-03': logWith('ex_9', [{ set_num: 1, weight: 65, reps: 6 }]),
    };
    const r = findLastPerformance('squat', '2026-06-07', plans, logs);
    assert.equal(r.date, '2026-06-03');
    assert.equal(r.sets[0].weight, 65);
});

test('planned-but-not-logged date is skipped -> next logged one', () => {
    const plans = {
        '2026-05-28': planWith('ex_1', 'squat'),
        '2026-06-04': planWith('ex_2', 'squat'),
    };
    const logs = {
        // 06-04: planned but only placeholder/empty sets -> not "performed"
        '2026-06-04': logWith('ex_2', [{ set_num: 1 }, { set_num: 2, completed: true }]),
        '2026-05-28': logWith('ex_1', [{ set_num: 1, weight: 40, reps: 12 }]),
    };
    const r = findLastPerformance('squat', '2026-06-07', plans, logs);
    assert.equal(r.date, '2026-05-28');
});

test('refDate excludes same-day and future sessions', () => {
    const plans = {
        '2026-06-07': planWith('ex_1', 'squat'),  // same day
        '2026-06-10': planWith('ex_2', 'squat'),  // future
        '2026-06-02': planWith('ex_3', 'squat'),  // prior
    };
    const logs = {
        '2026-06-07': logWith('ex_1', [{ set_num: 1, weight: 99, reps: 1 }]),
        '2026-06-10': logWith('ex_2', [{ set_num: 1, weight: 88, reps: 1 }]),
        '2026-06-02': logWith('ex_3', [{ set_num: 1, weight: 70, reps: 5 }]),
    };
    const r = findLastPerformance('squat', '2026-06-07', plans, logs);
    assert.equal(r.date, '2026-06-02');
});

test('only sets with data are returned (set_num preserved)', () => {
    const plans = { '2026-06-01': planWith('ex_1', 'squat') };
    const logs = { '2026-06-01': logWith('ex_1', [
        { set_num: 1 },                              // empty -> dropped
        { set_num: 2, weight: 60, reps: 8, rpe: 8 }, // kept
        { set_num: 3, weight: 60, reps: 7 },         // kept
    ]) };
    const r = findLastPerformance('squat', '2026-06-07', plans, logs);
    assert.deepEqual(r.sets.map(s => s.set_num), [2, 3]);
});

test('slug present in plan but no log at all -> skipped', () => {
    const plans = { '2026-06-01': planWith('ex_1', 'squat') };
    const r = findLastPerformance('squat', '2026-06-07', plans, {});
    assert.equal(r, null);
});

// ---- exposure-aware lookup -------------------------------------------------

test('no exposure requested -> most recent match regardless of exposures', () => {
    const plans = {
        '2026-06-02': planWith('ex_1', 'squat', 'HEAVY'),
        '2026-06-05': planWith('ex_2', 'squat', 'LIGHT'),
    };
    const logs = {
        '2026-06-02': logWith('ex_1', [{ set_num: 1, weight: 92.5, reps: 4 }]),
        '2026-06-05': logWith('ex_2', [{ set_num: 1, weight: 61, reps: 10 }]),
    };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs);
    assert.equal(r.date, '2026-06-05');
    assert.equal(r.sets[0].weight, 61);
});

test('same-exposure match wins over a NEWER other-exposure one', () => {
    const plans = {
        '2026-06-02': planWith('ex_1', 'squat', 'HEAVY'),
        '2026-06-05': planWith('ex_2', 'squat', 'LIGHT'),  // newer, wrong exposure
    };
    const logs = {
        '2026-06-02': logWith('ex_1', [{ set_num: 1, weight: 92.5, reps: 4 }]),
        '2026-06-05': logWith('ex_2', [{ set_num: 1, weight: 61, reps: 10 }]),
    };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r.date, '2026-06-02');
    assert.equal(r.sets[0].weight, 92.5);
});

test('older same-exposure sessions -> the most recent of that chain', () => {
    const plans = {
        '2026-05-19': planWith('ex_1', 'squat', 'HEAVY'),
        '2026-05-26': planWith('ex_2', 'squat', 'HEAVY'),
        '2026-06-05': planWith('ex_3', 'squat', 'LIGHT'),
    };
    const logs = {
        '2026-05-19': logWith('ex_1', [{ set_num: 1, weight: 85, reps: 5 }]),
        '2026-05-26': logWith('ex_2', [{ set_num: 1, weight: 87.5, reps: 5 }]),
        '2026-06-05': logWith('ex_3', [{ set_num: 1, weight: 61, reps: 10 }]),
    };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r.date, '2026-05-26');
});

test('no same-exposure history -> falls back to the newest any-exposure match', () => {
    // First cycle of a new exposure key: only the LIGHT chain exists.
    const plans = {
        '2026-05-29': planWith('ex_1', 'squat', 'LIGHT'),
        '2026-06-05': planWith('ex_2', 'squat', 'LIGHT'),
    };
    const logs = {
        '2026-05-29': logWith('ex_1', [{ set_num: 1, weight: 57.5, reps: 12 }]),
        '2026-06-05': logWith('ex_2', [{ set_num: 1, weight: 61, reps: 10 }]),
    };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r.date, '2026-06-05');
    assert.equal(r.sets[0].weight, 61);
});

test('a plan exercise with NO exposure never matches a requested one (but is fallback-eligible)', () => {
    // Pre-epoch history carries no exposure at all.
    const plans = { '2026-06-05': planWith('ex_1', 'squat') };
    const logs = { '2026-06-05': logWith('ex_1', [{ set_num: 1, weight: 80, reps: 6 }]) };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r.date, '2026-06-05');   // fallback, not a match

    // With a real HEAVY session present, the fallback loses to it even when older.
    plans['2026-05-22'] = planWith('ex_0', 'squat', 'HEAVY');
    logs['2026-05-22'] = logWith('ex_0', [{ set_num: 1, weight: 90, reps: 5 }]);
    const r2 = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r2.date, '2026-05-22');
});

test('an unlogged same-exposure session is skipped, not counted as the match', () => {
    const plans = {
        '2026-05-22': planWith('ex_1', 'squat', 'HEAVY'),
        '2026-06-01': planWith('ex_2', 'squat', 'HEAVY'),  // planned, empty sets
        '2026-06-05': planWith('ex_3', 'squat', 'LIGHT'),
    };
    const logs = {
        '2026-05-22': logWith('ex_1', [{ set_num: 1, weight: 90, reps: 5 }]),
        '2026-06-01': logWith('ex_2', [{ set_num: 1 }, { set_num: 2 }]),
        '2026-06-05': logWith('ex_3', [{ set_num: 1, weight: 61, reps: 10 }]),
    };
    const r = findLastPerformance('squat', '2026-06-09', plans, logs, 'HEAVY');
    assert.equal(r.date, '2026-05-22');
});

test('exposure requested but no history at all -> null', () => {
    assert.equal(findLastPerformance('squat', '2026-06-09', {}, {}, 'HEAVY'), null);
});

test('formatShortDate: YYYY-MM-DD -> "Mon D" without TZ shift', () => {
    assert.equal(formatShortDate('2026-06-01'), 'Jun 1');
    assert.equal(formatShortDate('2026-01-31'), 'Jan 31');
    assert.equal(formatShortDate('2026-12-09'), 'Dec 9');
    assert.equal(formatShortDate(''), '');
    assert.equal(formatShortDate(null), '');
});

File: android/core/data/src/test/kotlin/dev/jtiisto/wellness/core/data/coach/LastPerformanceTest.kt
package dev.jtiisto.wellness.core.data.coach

import dev.jtiisto.wellness.core.data.network.DateString
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.util.Locale

/**
 * The ghost-value lookup, transcribed case for case from
 * `test/js/last-performance.test.js` (17 cases).
 *
 * The fixtures keep the JS names and dates so a failure can be read straight
 * against the original.
 */
class LastPerformanceTest {

    @Test
    @DisplayName("setHasData: true only when a real metric is present")
    fun setHasDataOnlyForRealMetrics() {
        assertFalse(setHasData(null))
        assertFalse(setHasData(buildJsonObject { }))
        assertFalse(setHasData(loggedSet(setNum = 1)))
        assertFalse(setHasData(loggedSet(completed = true)))
        // 0 is real data: a zero-weight set is a bodyweight set.
        assertTrue(setHasData(loggedSet(weight = 0.0)))
        assertTrue(setHasData(loggedSet(reps = 5)))
        assertTrue(setHasData(loggedSet(rpe = 8.0)))
        assertTrue(setHasData(loggedSet(durationSec = 40)))
    }

    @Test
    @DisplayName("no history -> null")
    fun noHistory() {
        assertNull(findLastPerformance("squat", "2026-06-07", emptyMap(), emptyMap()))
    }

    @Test
    @DisplayName("missing/empty slug -> null")
    fun missingSlug() {
        val plans = mapOf("2026-06-01" to planWith("ex_1", "squat"))
        val logs = mapOf("2026-06-01" to logWith("ex_1", listOf(loggedSet(setNum = 1, reps = 5))))
        assertNull(findLastPerformance("", "2026-06-07", plans, logs))
        assertNull(findLastPerformance(null, "2026-06-07", plans, logs))
    }

    @Test
    @DisplayName("one prior in a different workout (different exercise_key) -> that session")
    fun priorInAnotherWorkout() {
        val plans = mapOf("2026-06-01" to planWith("ex_3", "squat"))
        val logs = mapOf(
            "2026-06-01" to logWith("ex_3", listOf(loggedSet(setNum = 1, weight = 60.0, reps = 8, rpe = 8.0))),
        )

        val result = findLastPerformance("squat", "2026-06-07", plans, logs)

        assertEquals("2026-06-01", result?.date)
        assertEquals(listOf(loggedSet(setNum = 1, weight = 60.0, reps = 8, rpe = 8.0)), result?.sets)
    }

    @Test
    @DisplayName("multiple priors -> most recent chosen")
    fun mostRecentPrior() {
        val plans = mapOf(
            "2026-05-20" to planWith("ex_1", "squat"),
            "2026-06-03" to planWith("ex_9", "squat"),
        )
        val logs = mapOf(
            "2026-05-20" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 50.0, reps = 10))),
            "2026-06-03" to logWith("ex_9", listOf(loggedSet(setNum = 1, weight = 65.0, reps = 6))),
        )

        val result = findLastPerformance("squat", "2026-06-07", plans, logs)

        assertEquals("2026-06-03", result?.date)
        assertEquals(65.0, result?.sets?.first()?.number("weight"))
    }

    @Test
    @DisplayName("planned-but-not-logged date is skipped -> next logged one")
    fun plannedButNotLoggedSkipped() {
        val plans = mapOf(
            "2026-05-28" to planWith("ex_1", "squat"),
            "2026-06-04" to planWith("ex_2", "squat"),
        )
        val logs = mapOf(
            // 06-04 was planned but only holds placeholders, so it was not performed.
            "2026-06-04" to logWith("ex_2", listOf(loggedSet(setNum = 1), loggedSet(setNum = 2, completed = true))),
            "2026-05-28" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 40.0, reps = 12))),
        )

        assertEquals("2026-05-28", findLastPerformance("squat", "2026-06-07", plans, logs)?.date)
    }

    @Test
    @DisplayName("refDate excludes same-day and future sessions")
    fun strictlyBeforeRefDate() {
        val plans = mapOf(
            "2026-06-07" to planWith("ex_1", "squat"),
            "2026-06-10" to planWith("ex_2", "squat"),
            "2026-06-02" to planWith("ex_3", "squat"),
        )
        val logs = mapOf(
            "2026-06-07" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 99.0, reps = 1))),
            "2026-06-10" to logWith("ex_2", listOf(loggedSet(setNum = 1, weight = 88.0, reps = 1))),
            "2026-06-02" to logWith("ex_3", listOf(loggedSet(setNum = 1, weight = 70.0, reps = 5))),
        )

        assertEquals("2026-06-02", findLastPerformance("squat", "2026-06-07", plans, logs)?.date)
    }

    @Test
    @DisplayName("only sets with data are returned (set_num preserved)")
    fun onlySetsWithData() {
        val plans = mapOf("2026-06-01" to planWith("ex_1", "squat"))
        val logs = mapOf(
            "2026-06-01" to logWith(
                "ex_1",
                listOf(
                    loggedSet(setNum = 1),
                    loggedSet(setNum = 2, weight = 60.0, reps = 8, rpe = 8.0),
                    loggedSet(setNum = 3, weight = 60.0, reps = 7),
                ),
            ),
        )

        val result = findLastPerformance("squat", "2026-06-07", plans, logs)

        assertEquals(listOf(2.0, 3.0), result?.sets?.map { it.number("set_num") })
    }

    @Test
    @DisplayName("slug present in plan but no log at all -> skipped")
    fun planWithoutLog() {
        val plans = mapOf("2026-06-01" to planWith("ex_1", "squat"))

        assertNull(findLastPerformance("squat", "2026-06-07", plans, emptyMap()))
    }

    // ---- exposure-aware lookup ---------------------------------------------

    @Test
    @DisplayName("no exposure requested -> most recent match regardless of exposures")
    fun noExposureRequested() {
        val result = findLastPerformance("squat", "2026-06-09", twoExposurePlans(), twoExposureLogs())

        assertEquals("2026-06-05", result?.date)
        assertEquals(61.0, result?.sets?.first()?.number("weight"))
    }

    @Test
    @DisplayName("same-exposure match wins over a NEWER other-exposure one")
    fun sameExposureBeatsNewer() {
        val result = findLastPerformance("squat", "2026-06-09", twoExposurePlans(), twoExposureLogs(), "HEAVY")

        assertEquals("2026-06-02", result?.date)
        assertEquals(92.5, result?.sets?.first()?.number("weight"))
    }

    @Test
    @DisplayName("older same-exposure sessions -> the most recent of that chain")
    fun mostRecentOfTheExposureChain() {
        val plans = mapOf(
            "2026-05-19" to planWith("ex_1", "squat", "HEAVY"),
            "2026-05-26" to planWith("ex_2", "squat", "HEAVY"),
            "2026-06-05" to planWith("ex_3", "squat", "LIGHT"),
        )
        val logs = mapOf(
            "2026-05-19" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 85.0, reps = 5))),
            "2026-05-26" to logWith("ex_2", listOf(loggedSet(setNum = 1, weight = 87.5, reps = 5))),
            "2026-06-05" to logWith("ex_3", listOf(loggedSet(setNum = 1, weight = 61.0, reps = 10))),
        )

        assertEquals("2026-05-26", findLastPerformance("squat", "2026-06-09", plans, logs, "HEAVY")?.date)
    }

    @Test
    @DisplayName("no same-exposure history -> falls back to the newest any-exposure match")
    fun fallsBackToNewestAnyExposure() {
        // First cycle of a new exposure key: only the LIGHT chain exists.
        val plans = mapOf(
            "2026-05-29" to planWith("ex_1", "squat", "LIGHT"),
            "2026-06-05" to planWith("ex_2", "squat", "LIGHT"),
        )
        val logs = mapOf(
            "2026-05-29" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 57.5, reps = 12))),
            "2026-06-05" to logWith("ex_2", listOf(loggedSet(setNum = 1, weight = 61.0, reps = 10))),
        )

        val result = findLastPerformance("squat", "2026-06-09", plans, logs, "HEAVY")

        assertEquals("2026-06-05", result?.date)
        assertEquals(61.0, result?.sets?.first()?.number("weight"))
    }

    @Test
    @DisplayName("a plan exercise with NO exposure never matches a requested one (but is fallback-eligible)")
    fun exposurelessPlanIsFallbackOnly() {
        // Pre-epoch history carries no exposure at all.
        val plans = mutableMapOf<DateString, PlanDto>("2026-06-05" to planWith("ex_1", "squat"))
        val logs = mutableMapOf<DateString, JsonObject>(
            "2026-06-05" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 80.0, reps = 6))),
        )

        assertEquals("2026-06-05", findLastPerformance("squat", "2026-06-09", plans, logs, "HEAVY")?.date)

        // With a real HEAVY session present, the fallback loses to it even when older.
        plans["2026-05-22"] = planWith("ex_0", "squat", "HEAVY")
        logs["2026-05-22"] = logWith("ex_0", listOf(loggedSet(setNum = 1, weight = 90.0, reps = 5)))

        assertEquals("2026-05-22", findLastPerformance("squat", "2026-06-09", plans, logs, "HEAVY")?.date)
    }

    @Test
    @DisplayName("an unlogged same-exposure session is skipped, not counted as the match")
    fun unloggedSameExposureSkipped() {
        val plans = mapOf(
            "2026-05-22" to planWith("ex_1", "squat", "HEAVY"),
            "2026-06-01" to planWith("ex_2", "squat", "HEAVY"),
            "2026-06-05" to planWith("ex_3", "squat", "LIGHT"),
        )
        val logs = mapOf(
            "2026-05-22" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 90.0, reps = 5))),
            "2026-06-01" to logWith("ex_2", listOf(loggedSet(setNum = 1), loggedSet(setNum = 2))),
            "2026-06-05" to logWith("ex_3", listOf(loggedSet(setNum = 1, weight = 61.0, reps = 10))),
        )

        assertEquals("2026-05-22", findLastPerformance("squat", "2026-06-09", plans, logs, "HEAVY")?.date)
    }

    @Test
    @DisplayName("exposure requested but no history at all -> null")
    fun exposureWithNoHistory() {
        assertNull(findLastPerformance("squat", "2026-06-09", emptyMap(), emptyMap(), "HEAVY"))
    }

    @Test
    @DisplayName("formatShortDate: YYYY-MM-DD -> \"Mon D\" without TZ shift")
    fun shortDate() {
        // Locale-pinned: the month name is localized here where the PWA indexes
        // a hardcoded English array (deviation 3).
        assertEquals("Jun 1", formatShortDate("2026-06-01", Locale.ENGLISH))
        assertEquals("Jan 31", formatShortDate("2026-01-31", Locale.ENGLISH))
        assertEquals("Dec 9", formatShortDate("2026-12-09", Locale.ENGLISH))
        assertEquals("", formatShortDate("", Locale.ENGLISH))
        assertEquals("", formatShortDate(null, Locale.ENGLISH))
        // Unparseable input is echoed rather than rendered as a wrong date.
        assertEquals("not-a-date", formatShortDate("not-a-date", Locale.ENGLISH))
    }

    private fun twoExposurePlans() = mapOf(
        "2026-06-02" to planWith("ex_1", "squat", "HEAVY"),
        "2026-06-05" to planWith("ex_2", "squat", "LIGHT"),
    )

    private fun twoExposureLogs() = mapOf(
        "2026-06-02" to logWith("ex_1", listOf(loggedSet(setNum = 1, weight = 92.5, reps = 4))),
        "2026-06-05" to logWith("ex_2", listOf(loggedSet(setNum = 1, weight = 61.0, reps = 10))),
    )
}
